### PR TITLE
Add ambiguous route analyzer

### LINKED ISSUE #45223: [Analyzer] Warn on ambiguous route patterns
## Background and Motivation

ASP.NET Core throws an error if a URL maps to two routes. This is fine, but it can cause confusion with new ASP.NET Core devs who are use to method overloading in C#. New developers might think that minimal or action endpoints also support overloading based on the bound type.

```csharp
app.MapGet("/product/{name}", (string name) => ...);
app.MapGet("/product/{id}", (int id) => ...);
```

For example, the routes above produces an ambiguous match because they have the same form from routing's perspective:
1. A literal `product` segment
2. An unconstrained route parameter segment.

It doesn't matter that the route parameters have different names or are bound to a minimal API method with different .NET types.

This limitation is easy to learn and fix. However, it's not obvious. The analyzer is aimed at new developers who I writing their first app. They can learn in the IDE instead of a runtime error.

## Proposed Analyzer

An analyzer that looks at routes and figures out whether they are ambiguous at coding time. This analyzer will only detect obvious and easily detectable situations. There are still many ways to get ambiguous route errors that the analyzer won't address.

### Analyzer Behavior and Message

The analyzer will only warn for ambiguous routes mapped in the same ~method~ (MapGet calls in a method could be conditional because they're in if statements) scope and using the same route builder. It won't attempt to detect matching routes across methods because they may be passed route builders created by `Group("...")`.

The analyzer message is the same message that ASP.NET Core throws at runtime.

#### Route handler analyzer

* Title: "Route conflict detected between route handlers"
* Message: "Route '{0}' conflicts with another handler route. An HTTP request that matches multiple routes results in an ambiguous match error. Fix the conflict by changing the route's path, HTTP method, or route constraints."

#### MVC route analyzer

* Title: "Route conflict detected between controller actions"
* Message: "Route '{0}' conflicts with another action route. An HTTP request that matches multiple routes results in an ambiguous match error. Fix the conflict by changing the route's path, HTTP method, or route constraints."

### Category

- [ ] Design
- [ ] Documentation
- [ ] Globalization
- [ ] Interoperability
- [ ] Maintainability
- [ ] Naming
- [ ] Performance
- [ ] Reliability
- [ ] Security
- [ ] Style
- [x] Usage

### Severity Level

- [ ] Error
- [x] Warning
- [ ] Info
- [ ] Hidden

Routing is complex. I (@JamesNK) feel more comfortable if we start this as a warning. If there are no false positives, then we can make it an error in the future.

## Usage Scenarios

```csharp
app.MapGet("/product/{name}", (string name) => ...);
app.MapGet("/product/{id}", (int id) => ...);
``` 

## Risks

False positives. When in doubt, the analyzer shouldn't warn.

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/45223

Not quite finished. The remaining work:
- [x] Check that the map call's `IEndpointConventionBuilder` isn't used to add metadata that could affect routing.
- [x] Allow list of metadata methods that don't affect matching and still allow endpoints to be compared, e.g. `RequireAuthorization`.

It only works with Minimal API. Most of the infrastructure is in place to expand it to analyze controller actions in the future.

### CODE REVIEW

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/RouteHandlerAnalyzer.cs
Comment (mavasani) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/RouteHandlerAnalyzer.cs:59]: Having said the above, it is probably unlikely we make the change of invoking Operation action callbacks within a single operation block in parallel as I think that would break many shipped analyzers. We already achieve parallelism by invoking callbacks for different operation blocks in parallel.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/RouteHandlerAnalyzer.cs
Comment (mavasani) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/RouteHandlerAnalyzer.cs:56]: This code will be executed in the OperationBlockStart callback itself - I presume you wanted it to be done at the end of OperationBlockEnd callback. I believe you want:

```csharp
context.RegisterOperationBlockEndAction(c =>
{
    DetectAmbiguousRoutes(c, mapOperations);

    mapOperations.Clear();
    concurrentQueue.Enqueue(mapOperations);
});
```

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/RouteHandlerAnalyzer.cs
Comment (mavasani) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/RouteHandlerAnalyzer.cs:94]: Where do you handle bailing out from block analysis if any invocation is within a conditional or loop statement?

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs
Comment (mavasani) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs:30]: Ah, so you are grouping invocations by common parenting block operation and analyzing each group. Will this handle invocations within conditional code that is not parented by a block operation? For example:

```csharp
void M(bool flag)
{
   InvocationToAnalyze();
   var result = flag ? InvocationToAnalyze() : InvocationToNotAnalyze();
}

```

Both `InvocationToAnalyze` calls above share the same parenting `IBlockOperation`, but it is not guaranteed that both calls will always be made on all code paths.

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DetectAmbiguousMappedRoutesTest.cs
Comment (mitchdenny) [inline src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DetectAmbiguousMappedRoutesTest.cs:90]: Clever test!

**Review by mitchdenny — APPROVED** [review]
Certainly a few edge cases I wouldn't have considered testing there!

**Comment by JamesNK** [comment]
Merging blocked until API review is approved.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs
Comment (Youssef1313) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs:18]: Already specified in the other partial declaration

```suggestion
public partial class RouteHandlerAnalyzer
```

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs
Comment (Youssef1313) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs:61]:
```suggestion
        if (current.Parent is IArgumentOperation { Parent: IInvocationOperation invocationOperation } &&
```

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs
Comment (Youssef1313) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs:146]: Stylistic suggestion

```suggestion
            return method.Name is "RequireRateLimiting" or "DisableRateLimiting";
```

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs
Comment (Youssef1313) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs:77]: Stylistic suggestion

```suggestion
            else if (current.Parent is IConditionalOperation or
                ICoalesceOperation or
                IAssignmentOperation or
                IArgumentOperation or
                IInvocationOperation)
```

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs
Comment (Youssef1313) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs:114]:
```suggestion
            return method.Name is "RequireAuthorization" or "AllowAnonymous";
```

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/AmbiguousRoutePatternComparer.cs
Comment (danmoseley) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/AmbiguousRoutePatternComparer.cs:33]:
```suggestion
                RoutePatternSegmentSeparatorNode _ => yPart is RoutePatternSegmentSeparatorNode,
```

Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/AmbiguousRoutePatternComparer.cs:33]: What changed here? I can't figure out what the difference is.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs
Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs:18]: I will leave it as is because the other partial route handler analyzers inherit each file. Cleaning up can happen in another PR.

**Comment by JamesNK** [comment]
API review approved, and all current feedback is applied.

Any other changes before merging?

**Pull request merged**
